### PR TITLE
Use async storage instead of localstorage

### PR DESCRIPTION
**Pull request by darkwing**
This may help with https://github.com/MetaMask/metamask-extension/issues/9821

Explanation:  

At present, we're setting and getting lots of data via localStorage, which is synchronous and can be slow.  This is an attempt at using indexedDB and make everything async.

### CODE REVIEW

File: ui/app/ducks/gas/gas.duck.js [hunk 1]
```diff
@@ -416,114 +420,113 @@ export function fetchGasEstimates(blockTime) {
...
-            .then((r) => r.json())
-            .then((r) => {
-              const estimatedPricesAndTimes = r.map(
-                ({ expectedTime, expectedWait, gasprice }) => ({
-                  expectedTime,
-                  expectedWait,
+    const shouldGetFreshGasQuote = Date.now() - timeLastRetrieved > 75000
+    let promiseToFetch = getStorageItem('GAS_API_ESTIMATES')
+    if (shouldGetFreshGasQuote) {
+      promiseToFetch = queryEthGasStationPredictionTable()
```
Comment (darkwing) [inline ui/app/ducks/gas/gas.duck.js:431]: I'll need to write all of this to remove the `then`s, etc.

File: ui/app/ducks/gas/gas.duck.js [hunk 2]
```diff
@@ -416,114 +420,111 @@ export function fetchGasEstimates(blockTime) {
...
+      estimates = priceAndTimeEstimates
+    }
+
+    dispatch(setPricesAndTimeEstimates(estimates))
+    dispatch(gasEstimatesLoadingFinished())
+    return estimates
   }
 }
 
 export function setCustomGasPriceForRetry(newPrice) {
```
Comment (Gudahtt) [inline ui/app/ducks/gas/gas.duck.js:524]: This is being called in `useRetryTransaction` in an async context, but without using `await`. Now that it's async, I think that hook should block until this is finished, so that the sidebar isn't shown before this completes.

File: ui/app/ducks/gas/gas.duck.js [hunk 3]
```diff
@@ -416,114 +420,111 @@ export function fetchGasEstimates(blockTime) {
...
+      ])
+
+      estimates = timeMappedToSeconds
+    } else if (priceAndTimeEstimates.length) {
+      estimates = priceAndTimeEstimates
+    }
+
+    dispatch(setPricesAndTimeEstimates(estimates))
+    dispatch(gasEstimatesLoadingFinished())
+    return estimates
```
Comment (Gudahtt) [inline ui/app/ducks/gas/gas.duck.js:520]: This used to return nothing beforehand, rather than `estimates`. Was this change intentional?

If this is removed, then line 429 above can be moved into a condition as well, so that `estimate` is only fetched in the case where it is used.

**Review by Gudahtt — COMMENTED** [review]
This looks fantastic! The new `storage-helper` API is much nicer.

Even if this doesn't help the related issue at all, using `IndexedDB` over Local Storage seems like an improvement, so that we aren't synchronously saving large objects.

**Comment by brad-decker** [comment]
Q: I'm wondering if we should move the new module into `/shared/modules` and replace the usages of local storage in the background process as well? Would this work in an extension's background process?

File: ui/app/ducks/gas/gas.duck.js (hunk 3 above)
Comment (darkwing) [reply ui/app/ducks/gas/gas.duck.js:520]: Lint complains `410:37  error  Expected to return a value at the end of async arrow function`; it seems logical to always return estimates from this function

Comment (Gudahtt) [reply ui/app/ducks/gas/gas.duck.js:520]: That error is from the `consistent-return` rule; it indicates that a value is returned in some branches but not others. It's being triggered by the `return null` early on in the function, and it can be addressed by returning `undefined` instead.

I'm not sure why we'd want to introduce a return value if it's never used :thinking:

Comment (Rojo123abc) [reply ui/app/ducks/gas/gas.duck.js:520]: Null

**Comment by Gudahtt** [comment]
@Rojo123abc please don't comment on our PRs unless you have something useful to contribute.

**Review by Rojo123abc — APPROVED** [review]

**Comment by darkwing** [comment]
> Q: I'm wondering if we should move the new module into `/shared/modules` and replace the usages of local storage in the background process as well? Would this work in an extension's background process? Not in this PR, though, this should land and then maybe consider elevating it to be shared.

I'll experiment with that.  It should work seamlessly, as `local-forage` falls back to localStorage when IndexedDB isn't available.

**Review by Gudahtt — APPROVED** [review]
LGTM!

We don't use the local storage API in the background, so this seems fine in just the UI for now. The background uses the [similar-but-different `storage.local` extension API](https://developer.chrome.com/extensions/storage), which is already async.
